dissect/target/plugin.py

```diff
@@ -540,74 +531,160 @@ def decorator(obj):
+        # The PLUGINS tree does not include the hierarchy up to the plugins
+        # directory (dissect.target.plugins) for the built-in plugins. For the
+        # plugins in the directory specified in --plugin-path, the hierarchy
+        # starts at that directory.
+        #
+        # Build-in OSPlugins do have the dissect.target.plugins path in their
+        # module name, so it needs to be stripped, e.g.:
+        #   dissect.target.plugins.general.default -> general.default
+        #   dissect.target.plugins.windows._os -> plugins.windows._os
+        #
+        # The module name of OSPlugins from --plugin-path starts at the
+        # directory specified in that option, e.g.:
+        # --plugin-path=/some/path/, with a file foo/baros/_os.py
+        # will have a module name of: foo.baros._os
+        filter_path = _modulepath(osfilter).split(".")
+
+        # If an OSPlugin is not defined in a file called _os.py, an extra `_os`
+        # part is added to the PLUGINS tree.
+        # For example the default OS plugin with module name general.default
+        # (after stripping of the build-in hierarchy) will be added at:
+        # general
+        #        \- default
+        #                  \- _os
+        # However the `_os` part is not in the module name. Modules that are
+        # defined in an _os.py file have the `_os` part in their module name.
+        # It is stripped out, so the filter is similar for both types of
+        # OSPlugin files.
+        if filter_path[-1] == "_os":
+            filter_path = filter_path[:-1]
+    else:
+        filter_path = []
 
-    def _walk(root=None):
+    def _walk(
+        root: dict,
+        special_keys: set[str] = set(),
+        only_special_keys: bool = False,
+        prev_module_path: list[str] = [],
+    ):
         for key, obj in root.items():
-            if key == special_key:
-                yield from obj
-
-            elif key.startswith("_"):
-                continue
-
-            elif "functions" not in obj:
-                yield from _walk(obj)
+            module_path = prev_module_path.copy()
+            module_path.append(key)
+
+            # A branch in the PLUGINS tree is traveresed to the next level if
+            # there are no filters (which in effect means all plugins are
+            # returned).
+            # If there is a filter, a branch is traversed if there is no `_os`
+            # item on the next level. If there there is an `_os` item, it is
+            # only traversed into if the module_path matches in full the
+            # (beginning of) the filter_path.
+            if (
+                not filter_path
+                or filter_path
+                and (
+                    issubclass(osfilter, general.default.DefaultPlugin)
+                    or "_os" not in obj
+                    or ("_os" in obj and (module_path == filter_path or module_path == filter_path[: len(module_path)]))
+                )
```

Comment:
```suggestion

            osfilter_is_a_default_plugin =  issubclass(osfilter, general.default.DefaultPlugin)
            no_os_in_obj =  "_os" not in obj
            os_full_match_begin = "_os" in obj and (module_path == filter_path or module_path == filter_path[: len(module_path)])

            if (not filter_path or filter_path and (osfilter_is_a_default_plugin or no_os_in_obj or os_full_match_begin)):
```

Given the complexity, maybe use intermediate variables?

Reply:
I made the statement less complex in a different way.